Add Management Cluster page with details on cluster updates (#823)
giantswarm/docs: src/content/general/management-clusters/index.md

---
linkTitle: Management Clusters
title: Giant Swarm Management Clusters
description: Describes the role Management Clusters perform in the Giant Swarm architecture and how they are updated.
menu:
  main:
    identifier: management-clusters
    parent: general
last_review_date: 2021-02-17
user_questions:
  - How do Giant Swarm use Management Clusters?
  - How are Management Clusters kept up to date?
owner:
  - https://github.com/orgs/giantswarm/teams/team-biscuit
---

## Overview of Management Clusters

> As we are fully convinced of Kubernetes as a platform for building platforms, we build all our management clusters based on Kubernetes. Giant Swarm leverages the concept of [“Operators"](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/) to control all resources that clusters need as [“Custom Resources”](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/).

Author: This is taken from the architecture docs but I think it works as an intro.

As well as Operators our management clusters run supporting components such as our monitoring stack and Happa our web UI. All components are deployed using the Giant Swarm [App Platform]({{< relref "/app-platform/overview" >}}).

## Workload Cluster releases

Our workload clusters are versioned using Giant Swarm releases. From a workload cluster point of view upgrades are described in [cluster upgrades]({{< relref "/general/cluster-upgrades" >}}).

For management clusters when we publish a new Giant Swarm release we create a new Release custom resource and any new operator versions are deployed to the management cluster.

> We deploy new instances of the operators with the new version to avoid impacting existing workload clusters. These new operators will not become active until existing workload clusters are upgraded or a new cluster is created.

Author: This to me is the key point. As its how we isolate MC updates from workload clusters.

Reviewer: This looks right to me

> We deploy new instances of the operators with the new version to avoid impacting existing workload clusters. These new operators will not become active until existing workload clusters are upgraded or a new workload cluster is created.

Author: This to me is the key point. As its how we isolate MC updates from workload clusters.

## Management cluster only components

> The components that are not part of a Giant Swarm release like our monitoring stack are deployed via app collections. We run the latest version of each component and the collection is managed by our legacy deployment component called draughtsman.

Author: I don't see a way of avoiding mentioning draughtsman here 😅

Reviewer: yeah, fine. maybe note that we're aiming to get rid of it this half?

Author: Sure I've added. 👍

> The components are deployed as App custom resources using our app platform. Changes are tested in a number of internal management clusters before being deployed to all management clusters.

Author: I think its best to be explicit here about the test process for MC changes but shout if that isn't helpful / needed.

Reviewer 2: Further above, `App Platform` has been title case. Please adjust for consistency.

## Configuration changes for Management Clusters

> TBC

Author: I think config changes are very important but I'd like to get feedback before going deeper.

I'm thinking of

- Management cluster config updates
- Workload cluster config updates
- Ensuring new versions of CRDs
- Rolling cluster certs

Reviewer: i'd probably break out ensuring new versions of CRDs / rolling cluster certs to separate issues, but sure. 

could you elucidate further on what WC config updates would mean?

Author: > i'd probably break out ensuring new versions of CRDs / rolling cluster certs to separate issues, but sure.

Good call I kept it to just config and I guess CRDs are config at a stretch. On rolling certs yes I think they are better treated separately and needs someone with more knowledge.

> could you elucidate further on what WC config updates would mean?

I'm thinking of changing catalog level config via `opsctl ensure appcatalogs` for workload cluster apps. We don't do this often but there is potential for things to go wrong.

I'll create a separate issue in batman for that and I think we need better docs for it.

Author: WC config issue for tracking https://github.com/giantswarm/giantswarm/issues/15796

## Management Cluster infrastructure upgrades

> TBC

Author: Do we need to mention Terraform updates here? 

If so I think these security docs explain it well. https://docs.giantswarm.io/security/operational-layers/

Reviewer: i think we _do_, even if we're planning on moving to using operators in future

Reviewer: maybe it's helpful to have a point towards that

Author: I've added it and included why we want to move to Operators and how they are better than Terraform and Kops.

@teemow I took ideas from https://www.giantswarm.io/management-cluster for that. Thanks I hadn't seen that page.

---

> linkTitle: Management Clusters

Reviewer 2: ```suggestion
linkTitle: Management clusters
```

> title: Giant Swarm Management Clusters

Reviewer 2: ```suggestion
title: Management Clusters
```

"Giant Swarm" is not needed in the title (because the context makes it self-evident that we talk about Giant Swarm stuff), but looks weird once it occurs in more and more titles/headlines.

Reviewer 2: Also changed to sentence case

description: The role Management Clusters perform in the Giant Swarm architecture and how they are updated.
weight: 60
menu:
  main:
    identifier: management-clusters
    parent: general
last_review_date: 2021-02-17
user_questions:
  - How do Giant Swarm use Management Clusters?
  - How are Management Clusters kept up to date?
owner:
  - https://github.com/orgs/giantswarm/teams/team-biscuit
---

> ## Overview of Management Clusters

Reviewer 2: ```suggestion
## Overview
```

> As we are fully convinced of Kubernetes as a platform for building platforms, we build all our management clusters based on Kubernetes. Giant Swarm leverages the concept of [“Operators"](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/) to control all resources that clusters need as [“Custom Resources”](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/).

Reviewer 3: IMO "operators" should be lowercase (elsewhere as well), but not sure if there's an official style rule for that

Reviewer 2: Yep, please use lowercase. Also, no quotes needed.

Author: Great I like that more too. Will fix and remove quotes.

> As well as operators, our management clusters run supporting components such as our monitoring stack and Happa, our web UI. All components are deployed using the Giant Swarm [App Platform]({{< relref "/app-platform/overview" >}}).

Reviewer 4: ```suggestion
As well as operators, our management clusters run supporting components such as our monitoring stack and our web UI. All components are deployed using the Giant Swarm [App Platform]({{< relref "/app-platform/overview" >}}).
```

Reviewer 4: Towards getting rid of Happa from the documentation.

Reviewer 2: Bonus points for linking `web UI` to `/ui-api/web/` :)

> As well as Operators our management clusters run supporting components such as our monitoring stack and Happa our web UI. All components are deployed using the Giant Swarm [App Platform]({{< relref "/app-platform/overview" >}}).

Reviewer 3: ```suggestion
As well as Operators, our management clusters run supporting components such as our monitoring stack and Happa, our web UI. All components are deployed using the Giant Swarm [App Platform]({{< relref "/app-platform/overview" >}}).
```
(commas. Also potentially "operators")

## Workload Cluster releases

> Our workload clusters are versioned using Giant Swarm releases. From a workload cluster point of view upgrades are described in [cluster upgrades]({{< relref "/general/cluster-upgrades" >}}).

Reviewer 3: ```suggestion
Our workload clusters are versioned using [Giant Swarm releases](https://github.com/giantswarm/releases/). From a workload cluster point of view, upgrades are described in [cluster upgrades]({{< relref "/general/cluster-upgrades" >}}).
```

Author: Much better, thx!

> Our workload clusters are versioned using [Giant Swarm releases](https://github.com/giantswarm/releases/). From a workload cluster point of view, upgrades are described in [cluster upgrades]({{< relref "/general/cluster-upgrades" >}}).

Reviewer 2: ```suggestion
Our workload clusters are versioned using [workload cluster releases]({{< relref "/general/releases" >}}). From a workload cluster point of view, upgrades are described in [cluster upgrades]({{< relref "/general/cluster-upgrades" >}}).
```

> When we publish a new Giant Swarm release, we create a new Release custom resource and any new operator versions are deployed to the management cluster.

Reviewer 4: ```suggestion
When we publish a new Giant Swarm release, we create a new release custom resource and any new operator versions are deployed to the management cluster.
```

> For management clusters when we publish a new Giant Swarm release we create a new Release custom resource and any new operator versions are deployed to the management cluster.

Reviewer 3: ```suggestion
When we publish a new Giant Swarm release, we create a new Release custom resource and any new operator versions are deployed to the management cluster.
```

Author: Yes that is simpler and flows better.

We deploy new instances of the operators with the new version to avoid impacting existing workload clusters. These new operators will not become active until existing workload clusters are upgraded or a new cluster is created.

> ## Management cluster only components

Reviewer 3: ```suggestion
## Management cluster-only components
```

> The components that are not part of a workload cluster release -- like our monitoring stack -- are deployed via app collections. We run the latest version of each component and the collection is managed by our legacy deployment component called draughtsman (which we plan to replace in Q2 2021).

Reviewer 4: Replace or retire/deprecate?

Author: I changed to retire. Thanks I think that's better.

> The components that are not part of a Giant Swarm release like our monitoring stack are deployed via app collections. We run the latest version of each component and the collection is managed by our legacy deployment component called draughtsman (which we plan to replace in Q2 2021).

Reviewer 3: ```suggestion
The components that are not part of a Giant Swarm release -- like our monitoring stack -- are deployed via app collections. We run the latest version of each component and the collection is managed by our legacy deployment component called draughtsman (which we plan to replace in Q2 2021).
```
Can alternatively use commas

> The components that are not part of a Giant Swarm release -- like our monitoring stack -- are deployed via app collections. We run the latest version of each component and the collection is managed by our legacy deployment component called draughtsman (which we plan to replace in Q2 2021).

Reviewer 2: ```suggestion
The components that are not part of a workload cluster release -- like our monitoring stack -- are deployed via app collections. We run the latest version of each component and the collection is managed by our legacy deployment component called draughtsman (which we plan to replace in Q2 2021).
```

The components are deployed as App custom resources using our app platform. Changes are tested in a number of internal management clusters before being deployed to all management clusters.

> ## Configuration changes for Management Clusters

Reviewer 2: ```suggestion
## Configuration changes
```

Avoiding repetition

> From time to time we also need to update the configuration for management clusters. These changes are also tested on our internal management clusters before being applied.

Reviewer 3: ```suggestion
From time to time, we also need to update the configuration for management clusters. These changes are also tested on our internal management clusters before being applied.
```

> ## Management Cluster infrastructure upgrades

Reviewer 2: ```suggestion
## Infrastructure upgrades
```

The infrastructure including the management cluster nodes and supporting resources like bastion hosts are managed by our SRE team using Terraform.

> However, we plan to move more of this provisioning to Operators reusing the components that manage our workload clusters where possible. As they are more dynamic compared to Terraform and other tools such as Kops. Operators continuously reconcile the cluster to its desired state.

Reviewer 3: ```suggestion
However, we plan to move more of this provisioning to operators, reusing the components that manage our workload clusters where possible. As they are continuously reconcile the cluster to its desired state, operators are more dynamic than Terraform and other tools such as Kops.
```

Author: Thanks this is better but I think it should be  "As they continuously reconcile". I'll update.
